File: python/test/part2_test.py
from src.video_player import VideoPlayer


def test_create_playlist(capfd):
    player = VideoPlayer()
    player.create_playlist("my_PLAYlist")
    out, err = capfd.readouterr()
    assert "Successfully created new playlist: my_PLAYlist" in out


def test_create_existing_playlist(capfd):
    player = VideoPlayer()
    player.create_playlist("my_cool_playlist")
    player.create_playlist("my_COOL_PLAYLIST")
    out, err = capfd.readouterr()
    assert "Successfully created new playlist: my_cool_playlist" in out
    assert ("Cannot create playlist: A playlist with the same name already "
            "exists") in out


def test_add_to_playlist(capfd):
    player = VideoPlayer()
    player.create_playlist("my_COOL_playlist")
    player.add_to_playlist("my_cool_PLAYLIST", "amazing_cats_video_id")
    out, err = capfd.readouterr()
    assert "Successfully created new playlist: my_COOL_playlist" in out
    assert "Added video to my_cool_PLAYLIST: Amazing Cats" in out


def test_add_to_playlist_already_added(capfd):
    player = VideoPlayer()
    player.create_playlist("my_cool_playlist")
    player.add_to_playlist("my_cool_playlist", "amazing_cats_video_id")
    player.add_to_playlist("my_cool_playlist", "amazing_cats_video_id")
    out, err = capfd.readouterr()
    assert "Successfully created new playlist: my_cool_playlist" in out
    assert "Added video to my_cool_playlist: Amazing Cats" in out
    assert "Cannot add video to my_cool_playlist: Video already added" in out


def test_add_to_playlist_nonexistent_video(capfd):
    player = VideoPlayer()
    player.create_playlist("my_cool_playlist")
    player.add_to_playlist("my_cool_playlist", "amazing_cats_video_id")
    player.add_to_playlist("my_cool_playlist", "some_other_video_id")
    out, err = capfd.readouterr()
    assert "Successfully created new playlist: my_cool_playlist" in out
    assert "Cannot add video to my_cool_playlist: Video does not exist" in out


def test_add_to_playlist_nonexistent_playlist(capfd):
    player = VideoPlayer()
    player.add_to_playlist("another_playlist", "amazing_cats_video_id")
    out, err = capfd.readouterr()
    assert "Cannot add video to another_playlist: Playlist does not exist" in out


def test_show_all_playlists_no_playlists_exist(capfd):
    player = VideoPlayer()
    player.show_all_playlists()
    out, err = capfd.readouterr()
    assert "No playlists exist yet" in out


def test_show_all_playlists(capfd):
    player = VideoPlayer()
    player.create_playlist("my_cool_playLIST")
    player.create_playlist("anotheR_playlist")
    player.show_all_playlists()
    out, err = capfd.readouterr()
    assert "Showing all playlists:" in out
    assert "my_cool_playLIST" in out
    assert "anotheR_playlist" in out
    # Need to check sorting starting from 4th line
    assert all(out.splitlines()[i] <= out.splitlines()[i + 1]
               for i in range(3, len(out.splitlines()) - 1))


def test_show_playlist(capfd):
    player = VideoPlayer()
    player.create_playlist("my_cool_playlist")
    player.show_playlist("my_COOL_playlist")
    player.add_to_playlist("my_cool_playlist", "amazing_cats_video_id")
    player.show_playlist("my_COOL_playlist")
    out, err = capfd.readouterr()
    assert "Successfully created new playlist: my_cool_playlist" in out
    assert "Showing playlist: my_cool_playlist" in out
    assert "No videos here yet" in out
    assert "Added video to my_cool_playlist: Amazing Cats" in out
    assert "Showing playlist: my_COOL_playlist" in out
    assert "Amazing Cats (amazing_cats_video_id) [#cat #animal]" in out


def test_remove_from_playlist_then_re_add(capfd):
    player = VideoPlayer()
    player.create_playlist("MY_playlist")
    player.add_to_playlist("my_playlist", "amazing_cats_video_id")
    player.add_to_playlist("my_playlist", "life_at_google_video_id")
    player.remove_from_playlist("my_playlist", "amazing_cats_video_id")
    player.add_to_playlist("my_playlist", "amazing_cats_video_id")
    player.show_playlist("my_playLIST")
    out, err = capfd.readouterr()
    assert "Showing playlist: my_playLIST" in out
    assert "Life at Google (life_at_google_video_id) [#google #career]" in out
    assert "Amazing Cats (amazing_cats_video_id) [#cat #animal]" in out
    assert out.find("Life at Google") < out.find("Amazing Cats")


def test_show_playlist_nonexistent_playlist(capfd):
    player = VideoPlayer()
    player.show_playlist("another_playlist")
    out, err = capfd.readouterr()
    assert "Cannot show playlist another_playlist: Playlist does not exist" in out


def test_remove_from_playlist(capfd):
    player = VideoPlayer()
    player.create_playlist("my_cool_playlist")
    player.add_to_playlist("my_cool_playlist", "amazing_cats_video_id")
    player.remove_from_playlist("my_COOL_playlist", "amazing_cats_video_id")
    player.remove_from_playlist("my_cool_playlist", "amazing_cats_video_id")
    out, err = capfd.readouterr()
    assert "Successfully created new playlist: my_cool_playlist" in out
    assert "Added video to my_cool_playlist: Amazing Cats" in out
    assert "Removed video from my_COOL_playlist: Amazing Cats" in out
    assert "Cannot remove video from my_cool_playlist: Video is not in playlist" in out


def test_remove_from_playlist_video_is_not_in_playlist(capfd):
    player = VideoPlayer()
    player.create_playlist("my_cool_playlist")
    player.remove_from_playlist("my_cool_playlist", "amazing_cats_video_id")
    out, err = capfd.readouterr()
    assert "Successfully created new playlist: my_cool_playlist" in out
    assert "Cannot remove video from my_cool_playlist: Video is not in playlist" in out


def test_remove_from_playlist_nonexistent_video(capfd):
    player = VideoPlayer()
    player.create_playlist("my_cool_playlist")
    player.add_to_playlist("my_cool_playlist", "amazing_cats_video_id")
    player.remove_from_playlist("my_cool_playlist", "some_other_video_id")
    out, err = capfd.readouterr()
    assert "Successfully created new playlist: my_cool_playlist" in out
    assert "Added video to my_cool_playlist: Amazing Cats" in out
    assert "Cannot remove video from my_cool_playlist: Video does not exist" in out


def test_remove_from_playlist_nonexistent_playlist(capfd):
    player = VideoPlayer()
    player.remove_from_playlist("my_cool_playlist", "amazing_cats_video_id")
    out, err = capfd.readouterr()
    assert "Cannot remove video from my_cool_playlist: Playlist does not exist" in out


def test_clear_playlist(capfd):
    player = VideoPlayer()
    player.create_playlist("my_cool_playlist")
    player.add_to_playlist("my_cool_playlist", "amazing_cats_video_id")
    player.show_playlist("my_cool_playlist")
    player.clear_playlist("my_COOL_playlist")
    player.show_playlist("my_cool_playlist")
    out, err = capfd.readouterr()
    assert "Successfully created new playlist: my_cool_playlist" in out
    assert "Added video to my_cool_playlist: Amazing Cats" in out
    assert "Showing playlist: my_cool_playlist" in out
    assert "Amazing Cats (amazing_cats_video_id) [#cat #animal]" in out
    assert "Successfully removed all videos from my_COOL_playlist" in out
    assert "Showing playlist: my_cool_playlist" in out
    assert "No videos here yet" in out


def test_clear_playlist_nonexistent(capfd):
    player = VideoPlayer()
    player.clear_playlist("my_cool_playlist")
    out, err = capfd.readouterr()
    assert "Cannot clear playlist my_cool_playlist: Playlist does not exist" in out


def test_delete_playlist(capfd):
    player = VideoPlayer()
    player.create_playlist("my_cool_playlist")
    player.delete_playlist("my_cool_playlist")
    out, err = capfd.readouterr()
    assert "Successfully created new playlist: my_cool_playlist" in out
    assert "Deleted playlist: my_cool_playlist" in out


def test_delete_playlist_nonexistent(capfd):
    player = VideoPlayer()
    player.delete_playlist("my_cool_playlist")
    out, err = capfd.readouterr()
    assert "Cannot delete playlist my_cool_playlist: Playlist does not exist" in out

File: cpp/test/part2_test.cpp
#include <gmock/gmock.h>
#include <gtest/gtest.h>

#include "../src/videoplayer.h"
#include "../src/helper.h"

using ::testing::HasSubstr;
using ::testing::MatchesRegex;

TEST(Part2, createPlaylist) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.createPlaylist("mY_plaYList");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 1);
  EXPECT_THAT(commandOutput[0],
              HasSubstr("Successfully created new playlist: mY_plaYList"));
}

TEST(Part2, createExistingPlaylist) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.createPlaylist("mY_plaYList");
  videoPlayer.createPlaylist("MY_PLAYLIST");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 2);
  EXPECT_THAT(commandOutput[0],
              HasSubstr("Successfully created new playlist: mY_plaYList"));
  EXPECT_THAT(commandOutput[1], HasSubstr("Cannot create playlist: A playlist with the "
                                "same name already exists"));
}

TEST(Part2, addToPlaylist) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.createPlaylist("mY_plaYList");
  videoPlayer.addVideoToPlaylist("mY_playLIST", "amazing_cats_video_id");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 2);
  EXPECT_THAT(commandOutput[0],
              HasSubstr("Successfully created new playlist: mY_plaYList"));
  EXPECT_THAT(commandOutput[1], HasSubstr("Added video to mY_playLIST: Amazing Cats"));
}

TEST(Part2, addToPlaylistAlreadyAdded) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.createPlaylist("mY_plaYList");
  videoPlayer.addVideoToPlaylist("mY_plaYList", "amazing_cats_video_id");
  videoPlayer.addVideoToPlaylist("mY_plaYList", "amazing_cats_video_id");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 3);
  EXPECT_THAT(commandOutput[0],
              HasSubstr("Successfully created new playlist: mY_plaYList"));
  EXPECT_THAT(commandOutput[1], HasSubstr("Added video to mY_plaYList: Amazing Cats"));
  EXPECT_THAT(
      commandOutput[2],
      HasSubstr("Cannot add video to mY_plaYList: Video already added"));
}

TEST(Part2, addVideoToPlaylistNonExistentVideo) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.createPlaylist("mY_plaYList");
  videoPlayer.addVideoToPlaylist("mY_plaYList", "amazing_cats_video_id");
  videoPlayer.addVideoToPlaylist("mY_plaYList", "some_other_video_id");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 3);
  EXPECT_THAT(commandOutput[0],
              HasSubstr("Successfully created new playlist: mY_plaYList"));
  EXPECT_THAT(commandOutput[1], HasSubstr("Added video to mY_plaYList: Amazing Cats"));
  EXPECT_THAT(
      commandOutput[2],
      HasSubstr("Cannot add video to mY_plaYList: Video does not exist"));
}

TEST(Part2, addVideoToPlaylistNonExistent) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.addVideoToPlaylist("anotHER_playlist", "amazing_cats_video_id");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 1);
  EXPECT_THAT(
      commandOutput[0],
      HasSubstr(
          "Cannot add video to anotHER_playlist: Playlist does not exist"));
}

TEST(Part2, addVideoToPlaylistNonExistentNoPlaylistNoVideo) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.addVideoToPlaylist("anotHER_playlist", "video_does_not_exist");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 1);
  EXPECT_THAT(
      commandOutput[0],
      HasSubstr(
          "Cannot add video to anotHER_playlist: Playlist does not exist"));
}

TEST(Part2, showAllPlaylistsNoPlaylistsExist) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.showAllPlaylists();
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 1);
  EXPECT_THAT(commandOutput[0], HasSubstr("No playlists exist yet"));
}

TEST(Part2, showAllPlaylist) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.createPlaylist("mY_plaYList");
  videoPlayer.createPlaylist("anotHER_playlist");
  videoPlayer.showAllPlaylists();
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 5);
  EXPECT_THAT(commandOutput[2], HasSubstr("Showing all playlists:"));
  EXPECT_THAT(commandOutput[4], HasSubstr("mY_plaYList"));
  EXPECT_THAT(commandOutput[3], HasSubstr("anotHER_playlist"));
}

TEST(Part2, showPlaylist) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.createPlaylist("mY_plaYList");
  videoPlayer.showPlaylist("mY_plaYList");
  videoPlayer.addVideoToPlaylist("mY_plaYList", "amazing_cats_video_id");
  videoPlayer.showPlaylist("mY_plaYList");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 6);
  EXPECT_THAT(commandOutput[0],
              HasSubstr("Successfully created new playlist: mY_plaYList"));
  EXPECT_THAT(commandOutput[1], HasSubstr("Showing playlist: mY_plaYList"));
  EXPECT_THAT(commandOutput[2], HasSubstr("No videos here yet"));
  EXPECT_THAT(commandOutput[3], HasSubstr("Added video to mY_plaYList: Amazing Cats"));
  EXPECT_THAT(commandOutput[4], HasSubstr("Showing playlist: mY_plaYList"));
  EXPECT_THAT(commandOutput[5],
              HasSubstr("Amazing Cats (amazing_cats_video_id) [#cat #animal]"));
}

TEST(Part2, showPlaylistAfterRemoveAVideoFromPlaylistThenReAdd) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.createPlaylist("mY_plaYList");
  videoPlayer.addVideoToPlaylist("mY_plaYList", "amazing_cats_video_id");
  videoPlayer.addVideoToPlaylist("mY_plaYList", "life_at_google_video_id");
  videoPlayer.removeFromPlaylist("mY_plaYList", "amazing_cats_video_id");
  videoPlayer.addVideoToPlaylist("mY_plaYList", "amazing_cats_video_id");
  videoPlayer.showPlaylist("mY_plaYList");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 8);
  EXPECT_THAT(commandOutput[5], HasSubstr("Showing playlist: mY_plaYList"));
  EXPECT_THAT(commandOutput[6],
              HasSubstr("Life at Google (life_at_google_video_id) [#google #career]"));
  EXPECT_THAT(commandOutput[7],
              HasSubstr("Amazing Cats (amazing_cats_video_id) [#cat #animal]"));
}

TEST(Part2, showPlaylistNonExistent) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.showPlaylist("mY_plaYList");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 1);
  EXPECT_THAT(
      commandOutput[0],
      HasSubstr("Cannot show playlist mY_plaYList: Playlist does not exist"));
}

TEST(Part2, removeFromPlaylist) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.createPlaylist("mY_plaYList");
  videoPlayer.addVideoToPlaylist("mY_plaYList", "amazing_cats_video_id");
  videoPlayer.removeFromPlaylist("MY_playlist", "amazing_cats_video_id");
  videoPlayer.removeFromPlaylist("mY_plaYList", "amazing_cats_video_id");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 4);
  EXPECT_THAT(commandOutput[0],
              HasSubstr("Successfully created new playlist: mY_plaYList"));
  EXPECT_THAT(commandOutput[1], HasSubstr("Added video to mY_plaYList: Amazing Cats"));
  EXPECT_THAT(commandOutput[2],
              HasSubstr("Removed video from MY_playlist: Amazing Cats"));
  EXPECT_THAT(
      commandOutput[3],
      HasSubstr(
          "Cannot remove video from mY_plaYList: Video is not in playlist"));
}

TEST(Part2, removeFromPlaylistVideoNotInPlaylist) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.createPlaylist("mY_plaYList");
  videoPlayer.removeFromPlaylist("mY_plaYList", "amazing_cats_video_id");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 2);
  EXPECT_THAT(
      commandOutput[1],
      HasSubstr(
          "Cannot remove video from mY_plaYList: Video is not in playlist"));
}

TEST(Part2, removeFromPlaylistNonexistentVideo) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.createPlaylist("mY_plaYList");
  videoPlayer.addVideoToPlaylist("mY_plaYList", "amazing_cats_video_id");
  videoPlayer.removeFromPlaylist("mY_plaYList", "some_other_video_id");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 3);
  EXPECT_THAT(commandOutput[0],
              HasSubstr("Successfully created new playlist: mY_plaYList"));
  EXPECT_THAT(commandOutput[1], HasSubstr("Added video to mY_plaYList: Amazing Cats"));
  EXPECT_THAT(
      commandOutput[2],
      HasSubstr("Cannot remove video from mY_plaYList: Video does not exist"));
}

TEST(Part2, removeFromPlaylistNonexistentPlaylist) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.removeFromPlaylist("my_cool_playlist", "some_other_video_id");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 1);
  EXPECT_THAT(commandOutput[0], HasSubstr("Cannot remove video from my_cool_playlist: "
                                "Playlist does not exist"));
}

TEST(Part2, clearPlaylist) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.createPlaylist("mY_plaYList");
  videoPlayer.addVideoToPlaylist("mY_plaYList", "amazing_cats_video_id");
  videoPlayer.showPlaylist("mY_plaYList");
  videoPlayer.clearPlaylist("mY_plaYList");
  videoPlayer.showPlaylist("mY_plaYList");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 7);
  EXPECT_THAT(commandOutput[0],
              HasSubstr("Successfully created new playlist: mY_plaYList"));
  EXPECT_THAT(commandOutput[1], HasSubstr("Added video to mY_plaYList: Amazing Cats"));
  EXPECT_THAT(commandOutput[2], HasSubstr("Showing playlist: mY_plaYList"));
  EXPECT_THAT(commandOutput[3],
              HasSubstr("Amazing Cats (amazing_cats_video_id) [#cat #animal]"));
  EXPECT_THAT(commandOutput[4],
              HasSubstr("Successfully removed all videos from mY_plaYList"));
  EXPECT_THAT(commandOutput[5], HasSubstr("Showing playlist: mY_plaYList"));
  EXPECT_THAT(commandOutput[6], HasSubstr("No videos here yet"));
}

TEST(Part2, clearPlaylistNonexistent) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.clearPlaylist("mY_plaYList");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 1);
  EXPECT_THAT(
      commandOutput[0],
      HasSubstr("Cannot clear playlist mY_plaYList: Playlist does not exist"));
}

TEST(Part2, deletePlaylist) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.createPlaylist("mY_plaYList");
  videoPlayer.deletePlaylist("MY_PLAYLIST");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 2);
  EXPECT_THAT(commandOutput[0],
              HasSubstr("Successfully created new playlist: mY_plaYList"));
  EXPECT_THAT(commandOutput[1], HasSubstr("Deleted playlist: MY_PLAYLIST"));
}

TEST(Part2, deletePlaylistNonexistent) {
  VideoPlayer videoPlayer = VideoPlayer();
  testing::internal::CaptureStdout();
  videoPlayer.deletePlaylist("mY_plaYList");
  std::string output = testing::internal::GetCapturedStdout();
  std::vector<std::string> commandOutput = splitlines(output);
  ASSERT_EQ(commandOutput.size(), 1);
  EXPECT_THAT(
      commandOutput[0],
      HasSubstr("Cannot delete playlist mY_plaYList: Playlist does not exist"));
}
